From: Mathias Gibbens <gibmat@debian.org>
Description: Fix FTBFS with updated etcd
Forwarded: not-needed
diff --git a/store/etcd/etcd.go b/store/etcd/etcd.go
index dcbbdca..18f2224 100644
--- a/store/etcd/etcd.go
+++ b/store/etcd/etcd.go
@@ -10,7 +10,7 @@ import (
 	"strings"
 	"time"
 
-	etcd "github.com/coreos/etcd/client"
+	etcd "github.com/coreos/etcd/client/v2"
 	"github.com/docker/libkv"
 	"github.com/docker/libkv/store"
 )
